java/arcs/sdk/Encoding.kt

```diff
@@ -0,0 +1,165 @@
+        }
+    }
+
+    fun decodeText(): String = chomp(getInt(':')).utf8ToString()
+
+    fun decodeNum(): Double = upTo(':').utf8ToString().toDouble()
+
+    fun decodeBool(): Boolean = chomp(1).utf8ToString() == "1"
+
+    companion object {
+        fun decodeDictionary(bytes: ByteArray): Map<String, String> {
+            val decoder = StringDecoder(bytes)
+            val dict = mutableMapOf<String, String>()
+
+            var num = decoder.getInt(':')
+            while (num-- > 0) {
+                val klen = decoder.getInt(':')
+                val key = decoder.chomp(klen).utf8ToString()
+
+                val vlen = decoder.getInt(':')
+                val value = decoder.chomp(vlen).utf8ToString()
+
+                dict[key] = value
+            }
+
+            return dict
+        }
+    }
+}
+
+class StringEncoder(
+    private val buffers: MutableList<ByteArray> = mutableListOf(),
+    private var size: Int = 0
+) {
+    fun encodeDictionary(dict: Map<String, Any?>): StringEncoder {
+        addStr("${dict.size}:")
+        for ((key, value) in dict) {
+            addBytes("", key.stringToUtf8())
+            encodeValue(value)
+        }
+        return this
+    }
+
+    fun encodeList(list: List<Any>): StringEncoder {
+        addStr("${list.size}:")
+        list.forEach { encodeValue(it) }
+        return this
+    }
+
+    fun encodeValue(value: Any?) {
+        when (value) {
+            is String -> addBytes("T", value.stringToUtf8())
+            is Boolean -> addStr("B${if (value) 1 else 0}")
+            is Double -> addStr("N$value:")
+            is Map<*, *> -> {
+                @Suppress("UNCHECKED_CAST")
+                val se = StringEncoder().encodeDictionary(value as Map<String, Any?>)
+                addBytes("D", se.toByteArray())
+            }
+            is List<*> -> {
+                @Suppress("UNCHECKED_CAST")
+                val se = StringEncoder().encodeList(value as List<Any>)
+                addBytes("A", se.toByteArray())
+            }
+            else -> throw IllegalArgumentException("Unknown expression.")
+        }
+    }
+
+    fun encode(prefix: String, str: String) {
+        addBytes(prefix, str.stringToUtf8())
+        addStr("|")
+    }
+
+    fun encode(prefix: String, num: Double) {
+        addStr("$prefix$num:|")
+    }
+
+    fun encode(prefix: String, flag: Boolean) {
+        addStr("$prefix${if (flag) "1" else "0"}|")
+    }
+
+    private fun addStr(str: String) {
+        str.stringToUtf8().also {
+            buffers.add(it)
+            size += it.size
+        }
+    }
+
+    private fun addBytes(prefix: String, bytes: ByteArray) {
+        addStr("$prefix${bytes.size}:")
+        buffers.add(bytes)
+        size += bytes.size
+    }
+
+    fun toByteArray(): ByteArray {
+        val res = ByteArray(size)
+        populate(res)
+        return res
+    }
+
+    fun toNullTermByteArray(): NullTermByteArray {
```

Comment:
Similar to toByteArray - might be nice to be able to re-use a NullTermByteArray when possible in calling-code.

Reply:
Added TODO